ACPs/99-validatorsetmanager-contract/README.md

```diff
@@ -0,0 +1,264 @@
+
+On each validator set change, the P-Chain generates a Warp message to notify any onchain program tracking the validator set. Onchain programs must be able to interpret this message, so they can trigger the appropriate action. These notification messages have not been specified to date.
+
+Given these assumptions and the fact that most of the active blockchains on Avalanche mainnet are EVM-based, we propose defining a reference implementation for a Solidity smart contract that can:
+
+1. Hold relevant information about the current Subnet validator set, as well as historical information
+2. Send validator set updates to the P-Chain by generating `AdressedCall`s defined in ACP-77
+3. Correctly update the validator set by interpreting notification messages received from the P-Chain
+4. Be easily extended with custom “security modules” to implement any security model
+
+Having an audited and open-source reference implementation freely available will contribute to lowering the cost of launching Subnets on Avalanche.
+
+Once deployed, this contract will be used as the `address` in the `SetSubnetValidatorManagerTx`.
+
+## Specification
+
+### Storage
+
+The `ValidatorSetManager` needs to store all this data to operate correctly.
+
+**Note:** This is pseudo-code that might not be 100% syntactically correct.
+
+```solidity
+// The ID of the Subnet that uses this contract to manage its validator set
+bytes32 subnetID;
+
+// The current Subnet validator set (list of NodeIDs)
+bytes32[] subnetCurrentValidatorSet;
+
+// Subnet validator
+struct SubnetValidator {
+    // The Warp message ID corresponding to the RegisterSubnetValidatorTx that added the validator.
+    // The message ID is needed to submit subsequent SetSubnetValidatorWeightTx.
+    bytes32 messageID,
+    // The nonce of the latest validator weight update
+    uint64 nonce,
+    // The validator weight
+    uint64 weight,
+    // The timestamp the validator was added
+    uint64 addTimestamp
+}
+
+// The Subnet current validators
+// NodeID => SubnetValidator
+mapping(bytes32 => SubnetValidator) subnetCurrentValidators;
+
+// The total weight of the current Subnet validator set
+uint64 subnetTotalWeight;
+
+// Subnet validator pending registration
+struct SubnetRegisteringValidator {
+    // The validator weight
+    uint64 weight,
+    // The timestamp after which the signature is invalid
+    uint64 expiry,
+    // The Ed25519 signature over the concatenated bytes of [subnetID]+[nodeID]+[blsPublicKey]+[weight]+[balance]+[expiry]
+    bytes signature
+}
+
+// The Subnet validators that are pending registration (= waiting for addition notification from the P-Chain)
+// NodeID => registering validator
+mapping(bytes32 => SubnetRegisteringValidator) subnetRegisteringValidators;
+
+// Subnet validation status
+enum SubnetValidationStatus {
+    Registering,
+    Active,
+    Removing,
+    Over
+}
+
+// Subnet validation
+struct SubnetValidation {
```

Comment:
Didn't quite get this struct, why do we need to keep a struct for a list of validation periods? Not sure we need a historical list of validation period updates